api/envoy/config/route/v3/route_components.proto

```diff
@@ -1564,3 +1571,35 @@ message QueryParameterMatcher {
     bool present_match = 6;
   }
 }
+
+// HTTP Internal Redirect :ref:`architecture overview <arch_overview_internal_redirects>`.
+message InternalRedirectPolicy {
+  message Predicate {
+    google.protobuf.Any typed_config = 1;
+  }
+
+  // An internal redirect is not handled, unless the number of previous internal redirects that a
+  // downstream request has encountered is lower than this value.
+  // In the case where a downstream request is bounced among multiple routes by internal redirect,
+  // the first route that hits this threshold, or does not set :ref:`internal_redirect_policy
+  // <envoy_api_field_config.route.v3.RouteAction.internal_redirect_policy>`
+  // will pass the redirect back to downstream.
+  //
+  // If not specified, at most one redirect will be followed.
+  google.protobuf.UInt32Value max_internal_redirects = 1;
+
+  // Defines what upstream response codes are allowed to trigger internal redirect. If unspecified,
+  // only 302 will be treated as internal redirect.
+  // Only 301, 302, 303, 307 and 308 are valid values.
+  repeated uint32 redirect_response_codes = 2
+      [(validate.rules).repeated = {items {uint32 {lte: 308 gte: 301}}}];
+
+  // Specifies a list of predicates that are queried when an upstream response is deemed
+  // to trigger an internal redirect by all other criteria. Any predicate in the list can reject
+  // the redirect, causing the response to be proxied to downstream.
+  repeated Predicate predicates = 3;
+
+  // Allow internal redirect to follow a target URI with a different scheme than the value of
+  // x-forwarded-proto. Default behavior is to not follow such redirects.
+  bool allow_cross_scheme_redirect = 4;
```

Comment:
What's your plan here?
So the default code base today allows serving secure and insecure content over TLS, and insecure content over raw HTTP.
Either you're coding up a behavioral change which changes the default (doable, but why not leave it permissive and add a predicate if you want to not allow?) or you're making it possible to serve TLS content in the clear which I'd really prefer we not do.

Reply:
Ok looking at the code this looks intentional, and I'm not sure how we want to document a difference in the new behavior, i.e. calling out to folks as they migrate that if they want to preserve existing behavior, they need to set something new.  It's probably worth sorting out what we want to do in general, though you could bypass that conversation if you invert this and have disallow_cross_scheme_redirect to disallow serving http content over an https channel.

Reply:
I took the route of using disallow_cross_scheme_redirect to be more permissive.

We have an upstream that receives traffic both directly from end user and from the Envoy. The upstream issues redirect scheme based on the incoming scheme. We would like to have the Envoy -> upstream traffic to always be TLS to protect internal only URL, but it's perfectly fine for the end user to talk to the Envoy in plain text otherwise. So cross scheme redirect is really valuable here:

Admittedly we can make the upstream smarter by looking not only to the tls context but also to x-forwarded-proto header, but the upstream then need to know when to look at x-forwarded-proto and when to not. Since it receives internet traffic, anyone can set such header to confuse it.